Add List Command (Less than 30 sponsors, TTY)

jtmcg opened the pull request:
This closes the first onboarding [task](https://github.com/orgs/github/projects/14079/views/2?pane=issue&itemId=71297713) associated with the `sponsors list` command. A few notes on choices:

1. I've interpreted the "Less than 30 sponsors" criteria to mean we just want to limit how many sponsors are shown. Thus, to keep things simple, I've implemented a 30 sponsor limit in the query and no logic to check sponsor count
2. Due to the differences required in union types for Reflection vs String Manipulation queries, I've decided to remove the Reflection implementation for this and instead just use the String Manipulation method

**How to test**
1. Pull down the branch
2. Run `script/build`
3. Try the `bin/gh sponsors list <username>` command

**Screenshot of working example**
![Screenshot 2024-07-25 at 2 33 25 PM](https://github.com/user-attachments/assets/35421223-899c-4819-9a74-edad902f7cc0)

**What type of feedback I'm looking for**

Any and all! I'm still learning the patterns and codebase, so any feedback will help uncover any unknown unknowns while I continue to ramp up.

jtmcg committed:
Add new sponsors command

This lays down the initial architecture for the new 'sponsors' command. It includes the new command directory and a 'hello world' type test for the 'list' command, validating the plumbing through the rest of the CLI.

jtmcg commented on pkg/cmd/sponsors/list/list.go:
```diff
@@ -0,0 +1,36 @@
+			opts.Username = args[0]
+			opts.Sponsors = []string{"GitHub"}
+
+			if runF != nil {
+				return runF(opts)
+			}
```
As far as I can tell, `runF` is required for testing. What other use-cases might we include a non-nil `runF`?

williammartin replied:
Great question! There's no other reason that I am aware of. This is a bit of test-induced design damage with the primary purpose being as a `spy` for the flag and argument parsing. However, in practice the cost has been low so I haven't given much thought towards how we might address it.

jtmcg committed:
Merge branch 'cli:trunk' into list-sponsors

jtmcg committed:
Extract sponsor list command logic from command declaration

In preparation for adding api calls, I've extracted the core functionality of the sponsor list command into a listRun function. This also updates the required tests to prepare for this structure.

jtmcg committed:
Add scaffolding for SponsorsList GraphQL query

Adds files and basic query for getting a user's sponsors via the GraphQL api. The query uses the reflection technique. A future commit will implement the same functionality using the string manipulation technique.

jtmcg committed:
Add querySponsorsViaStringManipulation and test

I've added the ability to query the GraphQL API via string manipulation
as well as refactored the original tests for querying for sponsors to
run the whole suite against both methods. That way, as I build out the feature,
I don't need to duplicate test suites for each implementation.

Note, I'm not sure I like this pattern, abstracting out the function name in
the test declaration, for most use cases because the tests can be hard to follow.
However, it seems fine for this exercise as it is conducive to me focusing
on learning the patterns around each method of querying.

jtmcg committed:
Add test for error case

I'm skeptical about the need for this, as it wasn't failing when I wrote
it. Go error handling and compilation errors seem to disallow not handling
this use-case, so this test may not be necessary. Adding it anyway as an
artifact for that revelation

jtmcg committed:
Plumb graphql query into `gh sponsors list <username>` command

Calling `gh sponsors list <username>` now queries the graphql endpoint
and prints the resulting sponsor list without formating

Noteably, there is an issue with the plubming for the
querySponsorsViaReflection function that has to do with typing of the
variables passed into the query. I've left a comment above the function
explaining more.

jtmcg commented on pkg/cmd/sponsors/list/list_test.go:
```diff
@@ -0,0 +1,90 @@
+		wants []string
+	}{
+		{
+			name: "happy path",
+			opts: &ListOptions{
+				HttpClient: func() (*http.Client, error) {
+					r := &httpmock.Registry{}
+
+					r.Register(
+						httpmock.GraphQL(`query SponsorsList\b`),
+						httpmock.StringResponse(`{"data": {"user": {"sponsors": {"totalCount": 2, "nodes": [{"login": "mona"}, {"login": "lisa"}]}}}}`))
+
+					return &http.Client{Transport: r}, nil
```
🤔 I copied this pattern over from another test file and it gave me pause. This smells like over-testing to me... There's already graphql stubs in [api/queries_sponsors_test.go](api/queries_sponsors_test.go) - copy/pasted here, in fact. What my reflex was here was to look into how to mock out `api.GetSponsorsList()` in line 48 of [pkg/cmd/sponsors/list/list.go](pkg/cmd/sponsors/list/list.go):

> opts.Sponsors, err = api.GetSponsorsList(httpClient, "", opts.Username)

Here's my reasons:

1. It avoids duplicative GraphQL mocks
2. It reduces the scope of the test to the interfaces between the functions, making these tests orthogonal (as written, this test is technically also testing the functionality of `api.GetSponsorsList`, which feels weird)
3. It ensures that this test won't be impacted by changes to `api.GetSponsorsList` (we should be able to rely on type safety for interfaces, maintaining the "contract" that creates to protect them)

What are y'all's thoughts, @andyfeller and @williammartin?

andyfeller replied:
> I copied this pattern over from another test file and it gave me pause. This smells like over-testing to me...

I think touches on what Will will call an abstraction code smell as the logic responsible for providing the data might not be fake-able so tests like this have to resort to working at the HTTP layer.

williammartin replied:
😬 Oh I have opinions on this alright.

williammartin replied:
I think that your reflex is bang on. I believe that the previous maintainers tried to integrate most of each command together in tests to increase confidence. In my opinion however, it has come at the cost of maintening the test suite.

The tests become quite challenging to read with such significant set up in the table and the setup repeats itself at different layers. More so, for commands that _share_ API requests, instead of being clear and saying "we have something that plays that role, it has its own tests, and here we're just going to return a stubbed value" we have to repeat the API set up across every test.

I can talk more about this when I share my implementation. I don't want to eagerly share it though :)

jtmcg replied:
Oooo... I love the teaser, @williammartin. I think this gives me the go-ahead to go tinker with a mocking approach 👍 Not exactly sure how to do it in Golang, but definitely interested in it!

jtmcg replied:
See my tinkering in [this PR](https://github.com/jtmcg/gh-cli/pull/1/files)

jtmcg committed:
[WIP] Add formating to sponsors list output

Uses the tableprinter to print out the list of a users sponsors with the
SPONSORS header.

WIP notes: the function is currently working as expected (at least for the
happy path), but I'm getting a invalid memory address error for the tests.
I think it's messing up in the iostreams.Test() on line 115, though I
haven't been able to figure out what I've done wrong, yet

williammartin commented on api/queries_sponsors.go:
```diff
@@ -0,0 +1,90 @@
+
+import "net/http"
+
+type SponsorQuery struct {
```
There's a tradeoff in the following idea, but one that I feel generally positive about making.

I _think_ that we should represent GQL queries as inlined anonymous structs like so:

```go
var listSponsorsQuery struct {
    User struct {
        Sponsors struct {
            Nodes []struct {
                Organization struct {
                    Login string
                } `graphql:"... on Organization"`
                User struct {
                    Login string
                } `graphql:"... on User"`
            }
        } `graphql:"sponsors(first: 30)"`
    } `graphql:"user(login: $login)"`
}
```

Why? Because in my opinion a query like this is a self-contained structure. It's much easier to see the exact shape of the data request/response when nested like this, rather than trying to mentally hold the pieces together.

The tradeoff is that unlike TypeScript, anonymous structs in Go are absolutely disgusting to construct values for which _can_ make them a bit unpleasant to use in tests. However, I could also make an argument that not reusing the same type in the implementation and test is _actually a good thing_ because it makes it less likely that the same mistake in structure appears in both places.

williammartin replied:
^ I'm not totally sure whether you need the union type here for `User` and `Organization`. I used it, and you used it in your string query but you didn't use it in your reflection query.

jtmcg replied:
~~I totally agree with your point, here. I was having some typing issues with your suggested implementation, so I decided to punt that refactor for later. The code snippet you've provided should accelerate me to that refactor 🙏 I'll dive back into this a bit after satisfying the rest of the AC for the first issue~~

Thought more about this. See below

jtmcg replied:
> However, I could also make an argument that not reusing the same type in the implementation and test is actually a good thing because it makes it less likely that the same mistake in structure appears in both places.

🤔 a bit more about this and I might argue this is overkill, because you've got another place that does this check: the mocked data. GraphQL won't work correctly if the shape of the mocked data doesn't match the expected response interface, so really this means we have 3 things that must line up among these types we're checking. This feels like overkill, so I could argue that reusing this interface between the code and tests to avoid this unpleasant construction in tests is more valuable than having a 3rd error check.

williammartin replied:
It's definitely the less valuable part in comparison to seeing the data shape at one glance. To avoid the gross construction I would lean towards having the test server return stringified JSON directly. This also has a minor benefit of clearly documenting what we should get back over the wire from the server.

With regards the server type checking, yes, absolutely. I'm not really describing increased confidence in final correctness, I'm speaking more to the development process. Like, I can run a query, grab the real JSON from the server, drop it into the test server response and go from there.

williammartin replied:
It's all tradeoffs though, of course. I just think over time I've landed on this as the most comfortable for me.

jtmcg replied:
Also, it seems the union type is required:

> GraphQL: Selections can't be made directly on unions (see selections on Sponsor) (query SponsorsList.user.sponsors.nodes.login)

I think this means I can't reuse the struct for the Reflection method vs the String Manipulation method (or I haven't been able to figure out how to reuse it, yet).

It seems the String Manipulation method allows a clear way to define that `login` can come from an organization or a user and the response body doesn't care about either. However, I suspect that the Reflection method requires the response to include both the Organization and User fields in the response and then map them accordingly, meaning that the []Nodes{} structs are different:

**String Manipulation**
```golang
type Node struct {
    Login string `json:"login"`
}
// and later in the query:
                                 nodes {
					... on User {
						login
					}
					... on Organization {
						login
					}
				}
```

**Reflection**
```golang
type Node struct {
    Organization struct {
        Login string `json:"org_login" graphql:"... on Organization`
    }
    User struct {
        Login string `json:"user_login" graphql:"... on User"`
    }
}
``` 

My gut says that String Manipulation is the better choice, here, because it utilizes the graphql BE to do the mapping of Org and User logins to a single list, rather than doing it locally.

That said, I'm probably missing something 😅

williammartin commented on api/queries_sponsors.go:
```diff
@@ -0,0 +1,90 @@
+func querySponsorsViaReflection(client *Client, hostname string, username string) (SponsorQuery, error) {
+	query := SponsorQuery{}
+	variables := map[string]interface{}{
+		"username": username,
```
```suggestion
                "username": githubv4.String(username),
```

I think this is probably going to fix your error but since I don't know exactly what it is I'm not sure.

andyfeller commented on api/queries_sponsors.go:
```diff
@@ -0,0 +1,90 @@
+
+import "net/http"
+
+type SponsorQuery struct {
+	User User `json:"user" graphql:"user(login: $username)"`
+}
+
+type User struct {
+	Sponsors Sponsors `json:"sponsors" graphql:"sponsors(first: 30)"`
+}
+
+type Sponsors struct {
+	TotalCount int    `json:"totalCount"`
+	Nodes      []Node `json:"nodes"`
+}
+
+type Node struct {
+	Login string `json:"login"`
+}
```
In your comment below about query via reflection / field tags, you could look at the organization code for an example:

https://github.com/cli/cli/blob/8da9ee2437ee95a39085ede1006f22283526d7fc/api/queries_org.go#L11-L21

jtmcg replied:
AMAZING! I was having issues getting the graphql strings correct and had to resort to this implementation. I'll see if I can piece it together from here 🙏

williammartin commented on api/queries_sponsors.go:
```diff
@@ -0,0 +1,90 @@
+	return data.SponsorQuery, nil
+}
+
+func mapQueryToSponsorsList(queryResult SponsorQuery) []string {
+	sponsorsList := []string{}
+	for _, node := range queryResult.User.Sponsors.Nodes {
+		sponsorsList = append(sponsorsList, node.Login)
+	}
+	return sponsorsList
+}
```
```suggestion
func mapQueryToSponsorsList(queryResult SponsorQuery) []string {
	sponsorsList := make([]string, len(queryResult.User.Sponsors.Nodes))
	for i, node := range queryResult.User.Sponsors.Nodes {
		sponsorsList[i] = node.Login
	}
	return sponsorsList
}
```

Not a huge deal but since we already know the number of nodes we can construct a slice of the exact size rather than using `append` which might cause reallocations behind the scenes.

jtmcg replied:
Smart! I definitely need to improve my reflex to think about memory allocation... It's not something that you generally think about in TypeScript and Python 😅 I'd ask that you keep calling out these opportunities for me to help me build the muscle 💪

williammartin commented on api/queries_sponsors.go:
```diff
@@ -0,0 +1,90 @@
+package api
```
In my opinion, this should go in the `sponsors/list` package. Copying from a question about whether types should go into `api` or not, I wrote:

I think mostly the answer is that they shouldn’t appear in `api`. For the most part domain types should be self contained within their respective packages. I confirmed with Sam at some point that there were two lingering reasons for api containing `Issue` , `PullRequest` et al. One was that they were the first ones before a different pattern was decided and two they were shared and they hadn’t taken the time to find a better home for them.

I think I generally agree with this direction. The thing is, we’re not building a generalised GitHub client, we’re building queries that exist to support certain commands. This is particularly evident and relevant when it comes to the GQL struct reflection queries, where we’re not say, representing everything about a `User` in the query, we’re narrowly scoping the struct type to the thing we care about for that command.

jtmcg replied:
Ah, good insight. I think I generally agree as well and will move it 👍

williammartin commented on api/queries_sponsors_test.go:
```diff
@@ -0,0 +1,118 @@
+			name: "success",
+			httpStubs: func(reg *httpmock.Registry) {
+				reg.Register(
+					httpmock.GraphQL(`query SponsorsList`),
```
You've done a great job copying the existing patterns here. I would argue that the testing harness here is not sufficient. For example, if we change query variables to:

```
	variables := map[string]interface{}{
		"username": "lolnope",
	}
```

No test fails.

If we change the query to:

```
	query := `query SponsorsList($username: String!) {

	}`
```

No test fails.

---

It's possible to get away with this kind of missing behaviour because for the most part it's _really_ obvious when you do or don't have the data you expect in manual testing. Personally, I don't think that we should be content with this though 😬

williammartin replied:
More on other possibilities at another time, I'm just whetting your appetite about some sketchy stuff.

jtmcg replied:
Yeah, I agree. This test suite is in no-way complete. There's been a certain tension between my preferred TDD loop and trying to adopt existing patterns. I've been focusing on understanding the code and trying to understand the patterns, but do plan to make this more robust. Thanks for the callout

williammartin replied:
Ah sorry, my message was not clear at all. I meant the http registry test harness was lacking and the tests you copied from weren't covering that. Wasn't a criticism of your approach. 

Though when i think about it there may be a couple of helpers i forgot about. Need to check another time.

jtmcg replied:
Ah, for sure. In that regard, I've been trying to limit myself to adding pieces to the test harness as I discover a need for them. That way I'm being intentional about what's getting added. I think I'll continue this approach for now and 🤞 that the need for the missing pieces becomes clear

jtmcg replied:
Update: nothing else has illuminated itself through development, so now it's time to ponder this 🤔

jtmcg replied:
I think I've addressed this if you can take another look

williammartin replied:
Definitely an improvement. What happens if you change the query in the implementation to:

```
	query := `query SponsorsList($username: String!) {

	}`
```

?

Personally, I think that most of the tests in the CLI that take the form:

```
					httpmock.GraphQL(`query SponsorsList`),
```

Are perhaps not _quite_ what I would hope because they don't validate the query shape. What do you think?

williammartin commented on api/queries_sponsors_test.go:
```diff
@@ -0,0 +1,118 @@
+	}{
+		{
+			name:    "querySponsorsViaReflection",
+			queryFn: querySponsorsViaReflection,
```
This is a clever use of injecting functions into the tests and it's nice you found it.

williammartin replied:
Ideally, we wouldn't be testing private functions here. I suspect you're doing it for pragmatism as you try out the different approaches but I just want to call it out as something that I would be looking to avoid in real codebase. There's a few times where testing of private functions has impacted my ability to refactor code.

williammartin replied:
That said, I do see that you have tested `formatTTYOutput` and `mapQueryToSponsorsList` as well, so maybe this is your preferred approach right now?

williammartin commented on api/queries_sponsors_test.go:
```diff
@@ -0,0 +1,118 @@
+				client := newTestClient(http)
+
+				sponsorsList, err := queryFn.queryFn(client, "github.com", "octocat")
+				if (err != nil) != tt.expectError {
```
I'm absolutely sure this exists in a million places but I think it's too clever. I would much prefer to see:

```
if tt.expectError {
    require.Error(t, err)
}
```

Could be personal opinion.

williammartin commented on pkg/cmd/sponsors/list/list.go:
```diff
@@ -0,0 +1,74 @@
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "List sponsors for a user",
+		Args:  cobra.ExactArgs(1),
```
We have a validator `cmdutil.ExactArgs` which allows for a message to be provided here. This is nice because the default message is horribly unclear.

williammartin commented on pkg/cmd/sponsors/list/list.go:
```diff
@@ -0,0 +1,74 @@
+		return fmt.Errorf("could not create http client: %w", err)
+	}
+
+	opts.Sponsors, err = api.GetSponsorsList(httpClient, "", opts.Username)
```
I would prefer not to have `Sponsors` hanging off the `ListOptions` struct. I think that it's putting multiple responsibilities onto this struct. It already holds:
 * Dependencies
 * Command Line Arguments

And now we're hanging something else off it as well.

williammartin commented on pkg/cmd/sponsors/list/list.go:
```diff
@@ -0,0 +1,74 @@
+		table.EndRow()
+	}
+
+	err := table.Render()
```
If you're not going to wrap errors (`fmt.Errorf("failed formatting TTY output: %v", err)`) then you might as well do `return table.Render()`

williammartin commented on pkg/cmd/sponsors/list/list_test.go:
```diff
@@ -0,0 +1,127 @@
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
```
You'll need to construct iostreams here to avoid a panic I think.

williammartin commented on pkg/cmd/sponsors/list/list_test.go:
```diff
@@ -0,0 +1,127 @@
+			},
+			wants: []string{
+				"SPONSORS",
+				"------",
```
I think you mentioned the test was failing before. The underline is stylistic, it's not actually characters on a new line. The styling is handled by ANSI color codes which you can read more about here: https://jafrog.com/2013/11/23/colors-in-terminal.html

jtmcg replied:
It was panicking, so I wasn't even getting this check yet. I think your above comment may be the fix I need 👍

williammartin commented on pkg/cmd/sponsors/list/list_test.go:
```diff
@@ -0,0 +1,127 @@
+package list
```
A bit of personal preference here but I would like to see us generally moving towards using `list_test` packages so that we have to work through public interfaces as the default.

jtmcg replied:
Let me repeat back to you what I think this means:

1. The tests for a given package should be their own package, called `package_test`
2. The package that's being tested would then be imported and referenced by `package.*`
3. This enforces the interfaces that are being tested to be public by default, because if they weren't we wouldn't be able to call them in the test suite

If this is correct, then I guess my question is why default to public interfaces for everything being tested? I've generally been of the opinion that public interfaces should only be provided for what is required for other packages. In this case, that means that the `NewCmdList` is the only thing that's public, because it's the only thing needed outside of the `list` package functionality. All the additional functions I'm testing are only their own functions to make testing easier and code more readable, so why make them public by default?

williammartin reviewed (commented):
I left a lot of comments. We can definitely get sync time if it's not clear.

jtmcg committed:
Move queries for sponsors list command from api package to list package

This is to conform with the existing patterns in the repo

jtmcg committed:
Refactor graphQL interface to allow for shallow mocking

There was an over-testing pattern in the code in that we were testing the
functionality of the querying function in both queries_test and list_test.

This was occuring because the current pattern doesn't allow for easy
mocking of the getter function and thus necessitated mocking the http
response for each function. This refactor introduces a Getter function to
the ListOptions interface and provides the factory functions required for
both normal execution and testing of the getter.

This approach has simplified Test_listRun in list_test.go by allowing for
direct mocking of the getter without requiring mocking of the graphQl
response, resulting in a more orthogonal test suite and non-duplicative
testing.

jtmcg committed:
Improve mapQueryToSponsorsList memory utilization

A small improvement, but given the size of the list we need this function
to return is known, the function now instantiates a list of fixed size
and assigns to it directly rather than calling `append` and potentially
cause memory reallocations behind the scenes

jtmcg committed:
Remove redundant error handling in formatTTYOutput

jtmcg committed:
Clean up query functions for `gh sponsors list <username>`

1. Remove unused querySponsorsViaReflection function and tests
1. Rename querySponsorsViaStringManipulation to querySponsors

jtmcg committed:
Add enforcement of username variable to queries_test.go

Previously, the test suite wasn't validating that the response of a given
graphQL query was dependent on the username passed in. This remedies that
oversight.

jtmcg committed:
Add better messaging for empty sponsors list

jtmcg committed:
Differentiate TTY vs non-TTY output for no found sponsors

To handle non-TTY output, a response with no sponsors now returns an empty
output instead of the TTY output

jtmcg committed:
Merge pull request #2 from jtmcg/handle-no-sponsors-case

Add better messaging for empty sponsors list

jtmcg committed:
Add prompting to sponsors list command when no username is provided

Prompting allows a user to call `gh sponsors list` with no arguments and
then enter the username they would like to query via the terminal prompt

jtmcg committed:
Merge pull request #1 from jtmcg/mock-getter-implementation

Refactor graphQL interface to allow for shallow mocking

jtmcg committed:
Check env for disabled prompting

This allows the `gh sponsors list` command to respect the
`GH_PROMPT_DISABLED` env variable when no argument is provided to the
command

jtmcg committed:
Add --json flag to `gh sponsors list <username>` command

This allows machine readable, json formating of the output flags

jtmcg committed:
Merge pull request #3 from jtmcg/add-prompting

Add prompting to sponsors list command when no username is provided

jtmcg committed:
Merge pull request #4 from jtmcg/add-json-flag

Add --json flag to `gh sponsors list <username>` command

williammartin commented on pkg/cmd/sponsors/list/list.go:
```diff
@@ -0,0 +1,107 @@
+		},
+	}
+
+	cmd.Flags().BoolVar(&opts.JSONResponse, "json", false, "output as json")
```
This is an obvious solution for the only reasonable behaviour of the command but if you check the task again, I don't think it passes the acceptance criteria.

However, less so than being "right", the intention of this task was to introduce you to the `cmdutil.AddJSONFlags` function.

williammartin commented:
Thoughts on closing this out @jtmcg? Just looking to clean up our open PRs. Cheers!
